### PR TITLE
Add cumprod support for device mps

### PR DESCRIPTION
**Pull request by Mr4k**
Related to #77764

Add support for the cumprod operation (which in turn allows its gradient). This also allows us to compute the gradient of prod since it was blocked behind cumprod in the case where exactly one element of the tensor was 0.

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/UnaryOps.mm [hunk 1]
```diff
@@ -406,4 +407,48 @@ Tensor logit_mps(const Tensor& self, c10::optional<double> eps) {
...
+              ")");
+  if (!is_macos_13_or_newer()) {
+    TORCH_WARN_ONCE("torch.cumprod supported by MPS on MacOS 13+, please upgrade");
+    auto cpu_result = self.to(at::Device(kCPU)).cumprod(dim, dtype);
+    at::_copy_from_and_resize(cpu_result, result);
+    return;
+  }
+  auto input = dtype.has_value() ? self.to(dtype.value()) : self;
+
+  // issue #103810551: cumprod is horribly broken for int8, int16 and as chances for overflow is pretty high, cast to
```
Comment (Mr4k) [inline aten/src/ATen/native/mps/operations/UnaryOps.mm:431]: I'm not sure what `issue #103810551` is and where I want to check if this applies to cumprod as well as cumsum. So far I assumed it did but seeing the actual issue in whatever tracker it lies would help

File: aten/src/ATen/native/mps/operations/UnaryOps.mm [hunk 2]
```diff
@@ -393,7 +394,7 @@ Tensor logit_mps(const Tensor& self, c10::optional<double> eps) {
 
   mps::unary_op(input,
                 result,
-                "cumsum_out_mp" + std::to_string(dim),
+                "cumsum_out_mps" + std::to_string(dim),
```
Comment (Mr4k) [inline aten/src/ATen/native/mps/operations/UnaryOps.mm:397]: drive by spelling fix

File: test/test_mps.py [hunk 3]
```diff
@@ -3682,6 +3681,54 @@ def helper(dtype):
...
+            y = x.cumprod(-1)
+
+            self.assertEqual(y, cpu_y)
+
+        [helper(dtype) for dtype in [torch.float32, torch.int16, torch.int32, torch.uint8]]
+
+    @unittest.skipIf(product_version < 13.0, "Skipped on macOS 12")
+    def test_cumprod_backward(self):
+        # the gradient computation of cumprod has two different paths depending on whether or not multiple derivatives are to be computed
+        for multiple_derivatives in [False, True]:
```
Comment (Mr4k) [inline test/test_mps.py:3721]: I am going to double check later that I am really exercising both paths successfully here

**Comment by Mr4k** [comment]
@kulinseth it appears this might be a duplicate pr with something on your branch: https://github.com/kulinseth/pytorch/pull/419, happy to close this out if need be

File: test/test_mps.py [hunk 4]
```diff
@@ -5175,6 +5222,30 @@ def helper(shape, dtype=torch.float32):
         for dtype in [torch.float32, torch.int32, torch.int64, torch.bool]:
             helper((2, 3), dtype)
 
+    # torch.cumprod is not supported on mps until macOS 13
+    @unittest.skipIf(product_version < 13.0, "Skipped on macOS 12")
+    def test_prod_backward(self):
```
Comment (Mr4k) [inline test/test_mps.py:5227]: For some reason the `
Mac MPS / macos-12-py3-arm64-mps / test` suite is failing with `test_mps.py::TestNLLLoss::test_prod_backward FAILED` despite that this test should be skipped. Am I misunderstanding the `skipIf` functionality here?

File: aten/src/ATen/native/mps/operations/UnaryOps.mm (hunk 1 above)
Comment (Mr4k) [reply aten/src/ATen/native/mps/operations/UnaryOps.mm:431]: from the work done on https://github.com/kulinseth/pytorch/pull/419, it appears constraint does hold for both functions so I will merge them into a single one as is done in this pr

**Comment by Mr4k** [comment]
I'm not sure why macos-12-py3 tests are failing for cumprod grad. It's likely something silly I missed. However I noticed the tests seem to not actually running be on macos12. If you look at the `Print Runner OS/HW Info` step of the CI it says `kern.osproductversion: 13.2.1` which seems suspicious.

**Review by kulinseth — APPROVED** [review]
Looks good, thanks.

**Comment by Mr4k** [comment]
@kulinseth I finally triaged all the bugs working on this pr uncovered and it's now ready to go!

**Pull request closed without merging**